### PR TITLE
Add Type, Variable, Pointer & Declaration to .codegen.ast

### PR DESCRIPTION
**Pull request by bjodah**
This is a work-in-progress PR for bringing new ast nodes into `.codegen.ast`.

#### TODO
- [x] Figure out correct use of `_sympify` with `args` of `Basic` (`None` does not sympfiy).
    Using `nargs = (1, 2)` having `None` as second argument is handled by only storing the first arg.
- [x] Figure out how strings should be handled as arguments to Basic. e.g. Symbol uses `.name`.
    `Type` now subclasses `Symbol`
- [x] Determine sane default for type inference from SymPy types to types relevant for native languages. (emphasis on machine native types handled by the FPU).
- Implement printers for `Declaration` in:
    - [x] CCodePrinter
    - [x] FCodePrinter

#### Quite likely also in this PR (since it would reveal if the `Type` abstraction is powerful enough):
- [x] Implement printers for float & long double math functions (`sinf`, `sinl` & friends)

#### Possibly also in this PR (since it would reveal if the `Declaration` class is sufficiently versatile):
- arrays:
  - [ ] contiguous 1D
  - [ ] 2D (column major, row major)
  - [ ] N-dimensional
  - [ ] strided
  - [ ] non-contiguous (e.g.: accessor macros / double pointers in C, `operator()` -> `T&` in C++)
  - [ ] compile time vs. runtime determined shapes / strides
  - [ ] constness and aliasing (`const`/`restrict`/`__restrict__` in C/C++, `InOut` etc. in Fortran)

### CODE REVIEW

**Comment by asmeurer** [comment]
Do you want to add printers here or should that be done in a separate pull request?

**Comment by bjodah** [comment]
I will add printers in this PR

**Comment by bjodah** [comment]
This might actually be a good time for review, what do you think @asmeurer ?

File: sympy/codegen/ast.py [hunk 1]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+
+    The naming is a super-set of NumPy naming, see [1]_.
+
+    Arguments
+    ---------
+    name : str or Type
+        Either an explicit type: ``intc``, ``intp``, ``int8``, ``int16``,
+        ``int32``, ``int64``, ``uint8``, ``uint16``, ``uint32``, ``uint64,
+        float16``, ``float32``, ``float64``, ``complex64``, ``complex128``,
+        ``bool. Or only kind (precision decided by code-printer): ``integer``,
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: "only kind" is awkward wording.

Comment (asmeurer) [inline sympy/codegen/ast.py]: Should also include `int` and `float`, to match the Python names.

File: sympy/codegen/ast.py [hunk 2]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+        ``real`` or ``complex`` (where the latter two are of floating point type).
+        If a ``Type`` instance is given, the said instance is returned.
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Should probably also support the builtins `int`, `float`, `complex` like NumPy.

File: sympy/codegen/ast.py [hunk 3]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+
+    Examples
+    --------
+    >>> from sympy.codegen.ast import Type
+    >>> Type.from_expr(42).name
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: You should mention this classmethod in the text above.

File: sympy/codegen/ast.py [hunk 4]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+            'eps': 1.08420217248550443401e-19,  # LDBL_EPSILON
+            'dig': 18,  # LDBL_DIG
+            'decimal_dig': 21,  # LDBL_DECIMAL_DIG
+        }
+    }
+    default_limits['complex64'] = default_limits['float32']
+    default_limits['complex128'] = default_limits['float64']
+
+    default_precision_targets = {  # e.g.:
+        # 'real': 'float64',
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Is this supposed to be commented out?

File: sympy/codegen/ast.py [hunk 5]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+    ValueError: Casting gives a significantly different value.
+    >>> Type('float80').cast_check(v18)
+    0.123456789012345649
+
+    References
+    ----------
+    [1] https://docs.scipy.org/doc/numpy/user/basics.types.html
+
+    """
+    allowed_names = tuple('intc intp int8 int16 int32 int64 uint8 uint16 uint32'.split() +
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: This is not used.

Comment (asmeurer) [inline sympy/codegen/ast.py]: Maybe instead of checking in the constructor, we can just have a set of known types as constants in this file, like

```
intc = Type('intc')
intp = Type('intp')
...
```

File: sympy/codegen/ast.py [hunk 6]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+    def from_expr(cls, expr, symb=None):
+        """ Infers type from an expression or a ``Symbol``.
+
+        Parameters
+        ----------
+        expr : number, string or SymPy object
+            The typename will be deduced from type or properties. Default is 'real'
+            (e.g. when a string is given as expr).
+        symb : Symbol (optional)
+            If given, assumptions of ``symb`` has higher precedence than expr.
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: If I am reading this right, `symb` is not actually used, other than for assumptions. Perhaps we could use the new assumptions here.

File: sympy/codegen/ast.py [hunk 7]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+
+    @classmethod
+    def from_expr(cls, expr, symb=None):
+        """ Infers type from an expression or a ``Symbol``.
+
+        Parameters
+        ----------
+        expr : number, string or SymPy object
+            The typename will be deduced from type or properties. Default is 'real'
+            (e.g. when a string is given as expr).
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: This is unclear.

File: sympy/codegen/ast.py [hunk 8]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+        symb : Symbol (optional)
+            If given, assumptions of ``symb`` has higher precedence than expr.
+
+        Examples
+        --------
+        >>> from sympy.codegen.ast import Type
+        >>> Type.from_expr(2) == Type('integer')
+        True
+        >>> Type.from_expr('i') == Type('integer')
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: This example is misleading. What is the point of allowing string inputs?

**Review by asmeurer — COMMENTED** [review]
My review so far. Will try to finish it up later today or tomorrow.

File: sympy/codegen/ast.py [hunk 9]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+    @property
+    def type(self):
+        return self.args[1]
+
+    @property
+    def const(self):
+        return self.args[2]
+
+
+class Pointer(Basic):
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Should it be a subtype of Variable?

File: sympy/printing/ccode.py [hunk 10]
```diff
@@ -394,37 +516,128 @@ def _print_NegativeInfinity(self, expr):
...
+    def _print_loggamma(self, expr):
+        return self._print_math_func(expr)
+
+    def _print_gamma(self, expr):
+        return self._print_math_func(expr)
+
+    def _print_ceiling(self, expr):
+        return self._print_math_func(expr)
+
+    def _print_floor(self, expr):
```
Comment (asmeurer) [inline sympy/printing/ccode.py]: You can just assign these all: `_print_floor = _print_ceiling = ... = _print_math_func`.

**Comment by asmeurer** [comment]
Seems weird to have lots of flags in .args as booleans. Not sure how I feel about that.

File: sympy/printing/tests/test_fcode.py [hunk 11]
```diff
@@ -685,3 +686,29 @@ def test_fcode_For():
...
+    var3 = Variable(y, (), Type('bool'))
+    dcl3 = Declaration(var3)
+    check(dcl3, 'logical :: y')
+
+    check(float32, "real(4)")
+    check(float64, "real(8)")
+    check(Type('real'), "real(4)", precision=float32)
+    check(Type('real'), "real(4)", precision=6)
+    check(Type('real'), "real(8)", precision=float64)
+    check(Type('real'), "real(8)", precision=15)
```
Comment (isuruf) [inline sympy/printing/tests/test_fcode.py]: In SymPy, we use `precision` for binary precision and `dps` for decimal precision. It'd be good to keep that for consistency. See `Float.__new__` for an example.

Comment (bjodah) [inline sympy/printing/tests/test_fcode.py]: Thanks, agreed. I inherited the terminology from the old printers (but I never liked that printer option anyway) -- I will change to dps.

Comment (bjodah) [inline sympy/printing/tests/test_fcode.py]: OK, this is going to be trickier than anticipated. Since `precision` is already an option in the printers this may need a deprecation cycle? Not sure how to proceed.

Comment (bjodah) [inline sympy/printing/tests/test_fcode.py]: We could start raising `SymPyDeprecationWarning` when a number is passed as `precision`?

Comment (asmeurer) [inline sympy/printing/tests/test_fcode.py]: Actually it's usually just `prec` for binary precision. We only used the longer name `precision` in Float because we have to change the behavior (previously `prec` was decimal precision).  (side note, I see no deprecation issue was opened for this)

Comment (bjodah) [inline sympy/printing/tests/test_fcode.py]: So should we introduce a new printer setting kwarg `prec` (taking a `Type` instance) and deprecate the use of printer setting named `precision`? (the Type will carry the dps info)

File: sympy/codegen/ast.py [hunk 12]
```diff
@@ -399,3 +400,405 @@ def body(self):
...
+            dec_dig = lim(corresponding_float, 'decimal_dig')
+            caster = lambda x: (
+                Float(str(re(x).evalf(dec_dig)), dec_dig+3) +
+                Float(str(im(x).evalf(dec_dig)), dec_dig+3)*1j
+            )
+            if abs(re(val)) > _max or abs(im(val)) > _max:
+                raise ValueError("Maximum value exceeded for data type.")
+            if abs(re(val)) < _tiny or abs(im(val)) < _tiny:
+                raise ValueError("Minimum (absolute) value for data type bigger than new value.")
+            new_val = caster(re(val)) + 1j*caster(im(val))
```
Comment (bjodah) [inline sympy/codegen/ast.py]: Yes, my assumption here is that there is a 1:1 correspondence between any floating point representation and its complex counter part (occupying twice the number of bits)

Comment (asmeurer) [inline sympy/codegen/ast.py]: I mean it seems like your definition is wrong. Should it just be `new_val = caster(val)`?

**Comment by bjodah** [comment]
The upcoming version of glibc (2.26) will support __float128. I think this is a good test for the `Type` class, do you agree? (I am in the middle of reworking it, will hopefully have something functioning later during the weekend).

File: sympy/printing/fcode.py [hunk 13]
```diff
@@ -55,15 +57,25 @@ class FCodePrinter(CodePrinter):
...
+        complex128: ('complex(8)', None),
+        Type('integer'): ('integer', None),
+        Type('bool'): ('logical', None)
+    }
+
     _default_settings = {
         'order': None,
         'full_prec': 'auto',
-        'precision': 15,
+        'precision': 17,
```
Comment (asmeurer) [inline sympy/printing/fcode.py]: Why did you change this? Assumedly 17 was used so that you always get full precision.

Comment (bjodah) [inline sympy/printing/fcode.py]: I changed it from 15 to 17 to preserve precision. There are some issues with printing 17 digits, e.g. the user might find ```
In [2]: Float(2.1).evalf(17)
Out[2]: 2.1000000000000001
```
confusing

Comment (asmeurer) [inline sympy/printing/fcode.py]: But isn't this used to evaluate symbolic constants? Is 15 digits enough to give the full machine precision for any literal?

Comment (bjodah) [inline sympy/printing/fcode.py]: It used to be 15, now it's 17

Comment (asmeurer) [inline sympy/printing/fcode.py]: Oh I somehow read the diff backwards.

**Comment by bjodah** [comment]
OK, I believe this is ready for review.

File: sympy/codegen/ast.py [hunk 14]
```diff
@@ -151,7 +193,7 @@ def aug_assign(lhs, op, rhs):
         subclass these types are also supported.
 
     op : str
-        Operator (+, -, /, *, %).
+        Operator (+, -, /, \*, %).
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Why did you add this? 

Also, this sort of change should be failing the tests. I guess my use of `python -We:invalid` in test_travis.sh isn't working. Probably because the files are bytecompiled already when sympy is installed. I wonder if adding `-B` as suggested [here](https://github.com/sympy/sympy/pull/12028) would fix it.

Comment (asmeurer) [inline sympy/codegen/ast.py]: Or maybe we just need to use `-We:invalid` when we call setup.py.

Comment (asmeurer) [inline sympy/codegen/ast.py]: Work here https://github.com/sympy/sympy/pull/13083

File: sympy/codegen/ast.py [hunk 15]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+            rtol = 1e-15 if exp10 is None else 2.0*ten**(-exp10)
+
+        def tol(num):
+            return atol + rtol*abs(num)
+
+        new_val = self._cast_nocheck(value)
+        self._check(new_val)
+
+        delta = new_val - val
+        if abs(delta) > tol(val):  # rounding, e.g. int(3.5) != 3.5
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: I hope you don't run into [this issue](https://github.com/sympy/sympy/pull/12537) for high precision (one of these days I'll finish that PR).

File: sympy/codegen/ast.py [hunk 16]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+    >>> half_precision.cast_check(1e5)  # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+      ...
+    ValueError: Maximum value for data type smaller than new value.
+
+    """
+    __slots__ = ['name', 'bits', 'max', 'tiny', 'eps', 'dig', 'decimal_dig']
+
+    def _cast_nocheck(self, value):
+        return Float(str(_sympify(value).evalf(self.decimal_dig)), self.decimal_dig)
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: `_sympify` disallows strings? Is that what we want?

File: sympy/codegen/ast.py [hunk 17]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+float32 = FloatType(
+    'float32', 32,
+    max=Float('3.40282347e+38', precision=32+8),
+    tiny=Float('1.17549435e-38', precision=32+8),
+    eps=Float('1.1920929e-07', precision=32+8),
+    dig=6, decimal_dig=9
+)
+float64 = FloatType(
+    'float64', 64,
+    max=Float('1.79769313486231571e+308', precision=64+8),
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: This adds three more zeros to the end of the number:

```
>>> Float('1.79769313486231571e+308', precision=64+8)
1.79769313486231571000e+308
```

Is that right?

File: sympy/codegen/ast.py [hunk 18]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+    >>> i = Symbol('i', integer=True)
+    >>> v = Variable.deduced(i)
+    >>> v.type == integer
+    True
+
+    """
+
+    nargs = (2, 3)  # type is optional
+
+    def __new__(cls, symbol, attrs=None, type_=None):
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Why not just set `attrs=FiniteSet()` here (or `S.EmptySet`)?

File: sympy/codegen/ast.py [hunk 19]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+        args = (_sympify(symbol), FiniteSet() if attrs is None else FiniteSet(*attrs))
+        if type_ is not None:
+            if not isinstance(type_, Type):
+                raise NotImplementedError("Expected a Type as type_")
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: This is a confusing error message. How about "TypeError("type_ argument should be an instance of Type")` (which is maybe also confusing...)

File: sympy/codegen/ast.py [hunk 20]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+    >>> k_decl.variable.type == integer
+    True
+
+    """
+
+    nargs = (1, 2)
+
+    def __new__(cls, var, val=None, cast=False):
+        if not isinstance(var, Variable):
+            raise NotImplementedError("Expected a Variable instance as var")
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Same here (and should be `TypeError`, assumedly)

File: sympy/codegen/ast.py [hunk 21]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+    Defining fields are set in __slots__.
+    """
+
+    __slots__ = []
+
+    def __new__(cls, *args, **kwargs):
+        if len(args) == 1 and not kwargs and isinstance(args[0], cls):
+            return args[0]
+        args = args + tuple([kwargs[k] for k in cls.__slots__[len(args):]])
+        return Type.__xnew_cached_(cls, *args)
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Token

File: sympy/codegen/ast.py [hunk 22]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+
+    def __new_stage2__(cls, *args):
+        obj = Basic.__new__(cls)
+        for attr, arg in zip(cls.__slots__, args):
+            setattr(obj, attr, arg)
+        return obj
+
+    __xnew_cached_ = staticmethod(cacheit(__new_stage2__))
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: I think we don't need all this stage2 junk here (or even caching, for that matter). As far as I can tell, it's used in Symbol so that symbols are cached based on derived assumptions (as opposed to the assumptions passed in by the user).

File: sympy/codegen/ast.py [hunk 23]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        for attr in self.__slots__:
+            if getattr(self, attr) != getattr(other, attr):
+                return False
+        return True
+
+    def __hash__(self):
+        return hash(tuple([getattr(self, attr) for attr in self.__slots__]))
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Should define `_hashable_content` instead of `__hash__`.

File: sympy/codegen/ast.py [hunk 24]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+
+    def __repr__(self):
+        return "{0}({1})".format(self.__class__.__name__, ', '.join(
+            ['%s=%s' % (k, repr(getattr(self, k))) for k in self.__slots__]
+        ))
+
+    __str__ = __repr__
+    _sympystr = lambda self, printer: repr(self)
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Is this not the default anyway?

Comment (asmeurer) [inline sympy/codegen/ast.py]: I would do the reverse: put the printing logic in here (so that it can recursively use the printer), and define `__repr__` based on `_sympystr`.

File: sympy/codegen/ast.py [hunk 25]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+        return 0
+
+    @property
+    def max(self):
+        return 2**self.bits - 1
+
+
+class FloatType(Type):
+    """ Represents a floating point value.
+
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Document what all the attributes mean.

File: sympy/codegen/ast.py [hunk 26]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+    ...     eps=Float('0.0009765625'), dig=2, decimal_dig=4)
+    >>> half_precision.cast_check(1.0)
+    1.0
+    >>> half_precision.cast_check(1e5)  # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+      ...
+    ValueError: Maximum value for data type smaller than new value.
+
+    """
+    __slots__ = ['name', 'bits', 'max', 'tiny', 'eps', 'dig', 'decimal_dig']
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: What is `dig`? I'm assuming that means "digits", but how is it different from `decimal_dig`? And if that does mean that, I think `dps` is a more standard abbreviation used in SymPy for decimal digits.

File: sympy/codegen/ast.py [hunk 27]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+
+        def tol(num):
+            return atol + rtol*abs(num)
+
+        new_val = self._cast_nocheck(value)
+        self._check(new_val)
+
+        delta = new_val - val
+        if abs(delta) > tol(val):  # rounding, e.g. int(3.5) != 3.5
+            raise ValueError("Casting gives a significantly different value.")
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Do you have a test that this doesn't fail if the user provides more digits than necessary?

File: sympy/codegen/ast.py [hunk 28]
```diff
@@ -399,3 +448,515 @@ def body(self):
...
+
+    def __new_stage2__(cls, *args):
+        obj = Basic.__new__(cls)
+        for attr, arg in zip(cls.__slots__, args):
+            setattr(obj, attr, arg)
+        return obj
+
+    __xnew_cached_ = staticmethod(cacheit(__new_stage2__))
+
+    def __eq__(self, other):
```
Comment (asmeurer) [inline sympy/codegen/ast.py]: Add tests for the different types to make sure == and != work properly.

File: sympy/printing/ccode.py [hunk 29]
```diff
@@ -104,6 +123,17 @@ def get_math_macros():
     }
 
 
+def _as_macro_if_defined(meth):
```
Comment (asmeurer) [inline sympy/printing/ccode.py]: Add a small docstring here explaining what this does.

File: sympy/printing/codeprinter.py [hunk 30]
```diff
@@ -14,6 +16,32 @@
...
+    """ Decorator for methods printing statements (as opposed to expressions) """
+
+    @wraps(method)
+    def wrapper(self, expr):
+        if self._settings.get('enable_statements', True):
+            return method(self, expr)
+        else:
+            raise ValueError("The code-printer setting 'enable_statements' is False,\n"
+                             "but %s generates a statement in the chosen language." %
+                             type(expr))
```
Comment (asmeurer) [inline sympy/printing/codeprinter.py]: Maybe this is too pedantic, but expressions themselves are valid statements in many languages (such as Python).

File: sympy/printing/ccode.py [hunk 31]
```diff
@@ -393,38 +560,44 @@ def _print_NegativeInfinity(self, expr):
...
 
-    def _print_exp2(self, expr):
-        return '{0}exp2({1})'.format(self._ns, self._print(expr.args[0]))
+    @requires(headers={'math.h'}, libraries={'m'})
+    @_as_macro_if_defined
+    def _print_math_func(self, expr, nest=False):
+        known = self.known_functions[expr.__class__.__name__]
+        if not isinstance(known, str):
+            for cb, name in known:
+                if cb(*expr.args):
```
Comment (asmeurer) [inline sympy/printing/ccode.py]: What does this do? Is there an example of this?

File: sympy/printing/ccode.py [hunk 32]
```diff
@@ -393,38 +560,44 @@ def _print_NegativeInfinity(self, expr):
...
+                    known = name
+                    break
+            else:
+                raise ValueError("No matching printer")
+        suffix = self._get_func_suffix(real) if self._ns + known in self._prec_funcs else ''
+        if nest:
+            args = self._print(expr.args[0])
+            if len(expr.args) > 1:
+                args += ', %s' % self._print(expr.fromiter(expr.args[1:]))
```
Comment (asmeurer) [inline sympy/printing/ccode.py]: I didn't even know about fromiter. `expr.func` is more standard.

File: sympy/codegen/ast.py (hunk 14 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: It's a Sphinx thing (asterisk needs to be escaped or it will be a reference), `\*` is interpreted as `\\*` by both Python 2 and Python 3. I will change this to '\\*' though.

Comment (asmeurer) [inline sympy/codegen/ast.py]: In Python 3.6, backslash escapes that don't actually escape anything are a syntax warning, and are tests should fail when they encounter them (it was broken, but I fixed it in https://github.com/sympy/sympy/pull/13083). In a future version of Python it will be a syntax error. You can make the docstring a raw string to avoid the issue.

File: sympy/codegen/ast.py (hunk 15 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Yeah I saw that one, tests have been fine so far though.

File: sympy/codegen/ast.py (hunk 16 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: I was thinking that symbols, e.g. 'x', should be disallowed. But the user could of course pass in e.g. `'123.456'`, changing this to `sympify`

File: sympy/codegen/ast.py (hunk 17 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: I've changed the representation of FloatType so max and min etc are now exact.

File: sympy/codegen/ast.py (hunk 19 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Agreed, updated

File: sympy/codegen/ast.py (hunk 18 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Agreed, updated

File: sympy/codegen/ast.py (hunk 20 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Agreed, updated

File: sympy/codegen/ast.py (hunk 22 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Sure, updated

File: sympy/codegen/ast.py (hunk 23 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Looking into Basic it seems to me I need both. I set _hashable_content = __hash__ now.

File: sympy/printing/codeprinter.py (hunk 30 above)
Comment (bjodah) [inline sympy/printing/codeprinter.py]: That's true, but it's a decorator that will be used in respective printer (would we ever want to print an expression as a statement? -- only use case I can think of is in try/except blocks and then enable_statements will be `True` any way. I guess we could have an option called "only_expressions" or something instead? I find "enable_statements" to be clearer though (minus the fact that an expression could also be a statement).

File: sympy/printing/ccode.py (hunk 31 above)
Comment (bjodah) [inline sympy/printing/ccode.py]: values in the known_functions dict may be functions. (e.g. "Abs")

File: sympy/printing/ccode.py (hunk 32 above)
Comment (bjodah) [inline sympy/printing/ccode.py]: `expr.func` does not work here.

File: sympy/codegen/ast.py (hunk 24 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Changed it, repr() does not give a string that can be `eval`:ed any more though.

**Comment by bjodah** [comment]
I think this is ready for another round of review.

File: sympy/codegen/ast.py (hunk 27 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: Added.

File: sympy/codegen/ast.py (hunk 17 above)
Comment (asmeurer) [inline sympy/codegen/ast.py]: Nice I like that a lot better.

File: sympy/codegen/ast.py (hunk 23 above)
Comment (asmeurer) [inline sympy/codegen/ast.py]: You should just set

```py
def _hashable_content(self):
     return tuple([getattr(self, attr) for attr in self.__slots__])
```

Does that not work?

File: sympy/printing/codeprinter.py (hunk 30 above)
Comment (asmeurer) [inline sympy/printing/codeprinter.py]: Maybe both are misleading because neither expressions or statements are a subset of the other in general. Even in C, for instance, you can't really say an expression is a statement because you need a semicolon to make a statement. And that matters because if you have `C_CODE_TEMPLATE.format(expr=ccode(expr))` then it matters if `expr` should be printed as an expression or statement simply by the question of whether ccode should print a semicolon or not.

File: sympy/printing/ccode.py [hunk 33]
```diff
@@ -393,38 +568,44 @@ def _print_NegativeInfinity(self, expr):
...
-        return '{0}log1p({1})'.format(self._ns, self._print(expr.args[0]))
+    # tgamma was already covered by 'known_functions' dict
 
-    def _print_exp2(self, expr):
-        return '{0}exp2({1})'.format(self._ns, self._print(expr.args[0]))
+    @requires(headers={'math.h'}, libraries={'m'})
+    @_as_macro_if_defined
+    def _print_math_func(self, expr, nest=False):
+        known = self.known_functions[expr.__class__.__name__]
+        if not isinstance(known, str):
```
Comment (asmeurer) [inline sympy/printing/ccode.py]: Need to use `string_types` from sympy.core.compatibility for Python 2 `unicode` compatibility.

File: sympy/codegen/ast.py (hunk 23 above)
Comment (bjodah) [inline sympy/codegen/ast.py]: this is weird, using `./bin/test` it works, but when I used pytest I got:
```
...
  File "/home/bjorn/vc/sympy/sympy/printing/ccode.py", line 168, in C89CodePrinter
    integer: intc
TypeError: unhashable type: 'Type'
```

Comment (bjodah) [inline sympy/codegen/ast.py]: Correction: it seems to be a Python2 vs. Python3 thing rather

Comment (bjodah) [inline sympy/codegen/ast.py]: So in Python 3 one [needs to](https://docs.python.org/3.6/reference/datamodel.html#object.__hash__) set `__hash__` explicitly if one define `__eq__`.

File: sympy/printing/ccode.py (hunk 31 above)
Comment (asmeurer) [inline sympy/printing/ccode.py]: Is this new to this PR? Seems to me that anything that's not just a string should be implemented as a printer method.

File: sympy/codegen/ast.py (hunk 23 above)
Comment (asmeurer) [inline sympy/codegen/ast.py]: Oh right. I think you have to define `__hash__` and have it call `super(Type, self).__hash__`. There are other examples in the codebase.

File: sympy/printing/codeprinter.py (hunk 30 above)
Comment (bjodah) [inline sympy/printing/codeprinter.py]: When in doubt I think the printer should just assume that it is dealing with an expression (e.g `ccode(x**2)` shouldn't append a semicolon). An alternative would be to enable printing of statements in a subclass of the CodePrinter, but due to backward compatibility CodePrinter will need to continue to support printing statements.

File: sympy/printing/ccode.py (hunk 33 above)
Comment (bjodah) [inline sympy/printing/ccode.py]: Got it, thanks.

File: sympy/printing/ccode.py (hunk 31 above)
Comment (bjodah) [inline sympy/printing/ccode.py]: It has been supported for a while:
https://github.com/sympy/sympy/blob/1.1.1/sympy/printing/tests/test_ccode.py#L51

Comment (asmeurer) [inline sympy/printing/ccode.py]: I see. I'm not a fan of that API (why not just provide a single function?), but I guess we need to keep it.
